### DISCUSSION TITLE
Better Null Analysis - Assert.IsNotNull

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5236] [Language Ideas] [answered]
It would be really nice if certain methods like Assert.IsNotNull could impact nullability.

**Comment by yaakov-h** [comment, marked as answer]
They already can if they use the appropriate attributes.

**Reply by TonyValenti** [reply]
What attributes do they need to do?  I am using nunit so I'll need to open a ticket with them.

**Reply by alrz** [reply]
See https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/InternalUtilities/Debug.cs

**Reply by manfred-brands** [reply]
You have to add the nunit.analyzer NuGet package.

**Reply by jnm2** [reply]
@TonyValenti To add more info: NUnit's `Assert.IsNotNull` does not always throw an exception when it records a failure. Specifically, within an Assert.Multiple block, Assert methods don't throw. C# nullability annotations aren't capable of matching the actual behavior by acting differently depending on whether Assert.IsNotNull is inside Assert.Multiple or not. (I documented this in the XML document comments for every Assert method.)

That's why @manfred-brands put (IMO amazing) effort in to make up for this gap by adding diagnostic suppression to  https://www.nuget.org/packages/NUnit.Analyzers.
